### ISSUE TITLE
SortedSet confusion with keys that sort (but not compare) the same

### ISSUE THREAD

**Issue by maleadt** [JuliaCollections/DataStructures.jl#528]
I'm having some issues with `SortedSet`. Reading the docs, I came up with the following:

```julia
julia> struct Foobar
       foo::Int
       bar::Int
       end

julia> foo(foobar::Foobar) = foobar.foo
foo (generic function with 1 method)

julia> const ByFoo = Base.By(foo)
Base.Order.By{typeof(foo)}(foo)


julia> a = Foobar(1,2)
Foobar(1, 2)

julia> push!(set, a)
SortedSet(Foobar[Foobar(1, 2)],
Base.Order.By{typeof(foo)}(foo))

julia> b = Foobar(1, 3)
Foobar(1, 3)

julia> push!(set, b)
SortedSet(Foobar[Foobar(1, 3)],
Base.Order.By{typeof(foo)}(foo))
```

So when using this ordering, for which two objects `a` and `b` sort identically, the `SortedSet` overwrites `a` when inserting `b`... This seems counter-intuitive, and not mentioned in the docs?

Anyhow, on a second read-through I noticed some details I missed (but IIUC do not explain the above) about `Base.lt` and `DataStructures.eq`, so I tried tweaking my comparator to have a proper egality function:

```julia
julia> DataStructures.eq(::typeof(ByFoo), a, b) = a == b

julia> DataStructures.eq(ByFoo, a, b)
false

# keys are different!


julia> set = SortedSet{Foobar}(ByFoo)
SortedSet(Foobar[],
Base.Order.By{typeof(foo)}(foo))

julia> push!(set, a)
SortedSet(Foobar[Foobar(1, 2)],
Base.Order.By{typeof(foo)}(foo))

julia> push!(set, b)
SortedSet(Foobar[Foobar(1, 2), Foobar(1, 3)],
Base.Order.By{typeof(foo)}(foo))

julia> # yay!

julia> haskey(set, b)
true

julia> haskey(set, a)
false

julia> # booo

julia> delete!(set, a)
ERROR: KeyError: key Foobar(1, 2) not found
```

This seems even worse, especially I can actually _get_ `a` and `b` by iterating the set!

Did I confuse `SortedSet` by defining an invalid comparator? Either way, maybe I've missed something but I think the docs could use some elaboration here.

**Comment by StephenVavasis** [comment]
I can try to improve the docs on this topic, but I'm not sure how.  I see some of the docs in-line with the code, other docs in the github directory docs/src, and then there is also documentation on github.io linked from the data structures home page that has a link labeled "edit on github".  Could someone who understands how the docs are laid out explain the correct way to create a PR for documentation corrections?

**Comment by StephenVavasis** [comment]
Closed via #787
